### PR TITLE
Convert MockWebServer to run on Okio.

### LINKED ISSUE #1175: Update MockWebServer to run on Okio

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #1175.

### CODE REVIEW

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 1]
```diff
@@ -531,30 +523,42 @@ private RecordedRequest readRequest(Socket socket, InputStream in, OutputStream
...
     boolean hasBody = false;
-    TruncatingOutputStream requestBody = new TruncatingOutputStream();
+    Buffer buffer = new Buffer();
+    TruncatingSink requestBody = new TruncatingSink(buffer);
     List<Integer> chunkSizes = new ArrayList<>();
-    MockResponse throttlePolicy = dispatcher.peek();
     if (contentLength != -1) {
-      hasBody = contentLength > 0;
-      throttledTransfer(throttlePolicy, socket, in, requestBody, contentLength);
+      if (contentLength > 0) {
```
Comment (JakeWharton) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: Going to leave this logic as-is because I like it better anyway.

**Comment by JakeWharton** [comment]
More work done here. Still not passing all tests.

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 2]
```diff
@@ -593,60 +598,63 @@ private void sleepIfDelayed(MockResponse response) {
...
+    public ThrottledSink(Sink delegate, MockResponse throttlePolicy) {
+      super(delegate);
+      bytesPerPeriod = throttlePolicy.getThrottleBytesPerPeriod();
+      delayMs = throttlePolicy.getThrottlePeriod();
+      periodRemainingBytes = bytesPerPeriod;
+    }
+
+    @Override public void write(Buffer source, long byteCount) throws IOException {
+      while (byteCount > 0) {
+        long actualCount = Math.min(Math.min(bytesPerPeriod, byteCount), periodRemainingBytes);
```
Comment (swankjesse) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: isn't `bytesPerPeriod` always ≤ `periodRemainingBytes` ?

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 3]
```diff
@@ -762,36 +754,18 @@ private void writeResponse(SpdyStream stream, MockResponse response) throws IOEx
...
-              long delayMs = response.getThrottleUnit().toMillis(response.getThrottlePeriod());
-              if (delayMs != 0) Thread.sleep(delayMs);
-            } catch (InterruptedException e) {
-              throw new AssertionError();
-            }
-          }
-        }
+      if (hasBody) {
+        sleepIfDelayed(response);
+        BufferedSink sink = Okio.buffer(new ThrottledSink(stream.getSink(), response));
```
Comment (swankjesse) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: so much nicer

File: okhttp-tests/src/test/java/com/squareup/okhttp/CacheTest.java [hunk 4]
```diff
@@ -840,8 +836,7 @@ private void testMethodInvalidates(String requestMethod) throws Exception {
         .addHeader("Last-Modified: " + lastModifiedDate)
         .addHeader("Expires: " + formatDate(1, TimeUnit.HOURS))
         .addHeader("Cache-Control: no-cache"));
-    List<String> headers = conditionalRequest.getHeaders();
-    assertTrue(headers.contains("If-Modified-Since: " + lastModifiedDate));
+    assertEquals(lastModifiedDate, conditionalRequest.getHeaders().get("If-Modified-Since"));
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/CacheTest.java]: much win

File: okhttp/src/main/java/com/squareup/okhttp/Headers.java [hunk 5]
```diff
@@ -130,6 +131,34 @@ public Builder newBuilder() {
...
+  @Override public int hashCode() {
+    int result = 0;
+    for (int i = 0, size = namesAndValues.length; i < size; i += 2) {
+      result = result * 37 + namesAndValues[i].toLowerCase(Locale.US).hashCode();
+      result = result * 37 + namesAndValues[i + 1].hashCode();
+    }
+    return result;
+  }
+
+  @Override public boolean equals(Object obj) {
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/Headers.java]: Document that we require the same headers in the same order?

**Comment by swankjesse** [comment]
LGTM

**Comment by JakeWharton** [comment]
Since this is spiraling into a bigger change than I wanted (since API changes have to happen) I'm going to break it up into multiple PRs.

**Pull request closed without merging**